x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealm.java

```diff
@@ -418,27 +476,114 @@ public void authenticate(final AuthenticationToken authenticationToken, final Ac
+                    listener.onResponse(AuthenticationResult.unsuccessful(msg, null));
+                    return;
+                }
+
+                JwtValidateUtil.validateType(jwt);
+                JwtValidateUtil.validateIssuer(jwt, allowedIssuer);
+                JwtValidateUtil.validateAudiences(jwt, allowedAudiences);
+                JwtValidateUtil.validateSignatureAlgorithm(jwt, isJwtSigPkc ? this.allowedJwksAlgsPkc : this.allowedJwksAlgsHmac);
+                JwtValidateUtil.validateAuthTime(jwt, now, this.allowedClockSkew.seconds());
+                JwtValidateUtil.validateIssuedAtTime(jwt, now, this.allowedClockSkew.seconds());
+                JwtValidateUtil.validateNotBeforeTime(jwt, now, this.allowedClockSkew.seconds());
+                JwtValidateUtil.validateExpiredTime(jwt, now, this.allowedClockSkew.seconds());
+
+                // At this point, client authc and JWT kty+alg+iss+aud+time filters passed. Do sig last, in case JWK reload is expensive.
+                try {
+                    JwtValidateUtil.validateSignature(
+                        jwt,
+                        isJwtSigPkc ? this.contentAndJwksAlgsPkc.jwksAlgs.jwks : this.contentAndJwksAlgsHmac.jwksAlgs.jwks
+                    );
+                } catch (Exception originalValidateSignatureException) {
+                    if (isJwtSigPkc) {
+                        final String sigErr = originalValidateSignatureException.getMessage() + " ";
+                        LOGGER.trace(sigErr + "Reloading PKC JWKs to retry verify JWT token=[" + tokenPrincipal + "]");
+                        final ContentAndJwksAlgs newContentAndJwksAlgs;
+                        try {
+                            newContentAndJwksAlgs = this.parseJwksAlgsPkc();
+                        } catch (Exception reloadException) {
+                            final String msg = sigErr + "Failed to reload PKC JWKs, can't retry verify JWT token=[" + tokenPrincipal + "]";
+                            reloadException.addSuppressed(originalValidateSignatureException);
+                            LOGGER.error(msg, reloadException);
+                            listener.onResponse(AuthenticationResult.unsuccessful(msg, reloadException));
+                            return;
+                        }
+                        if (Objects.equals(this.contentAndJwksAlgsPkc.jwksContent, newContentAndJwksAlgs.jwksContent)) {
+                            final String msg = sigErr + "Reloaded same PKC JWKs, can't retry verify JWT token=[" + tokenPrincipal + "]";
+                            final ElasticsearchException reloadPkcException = new ElasticsearchException(msg);
+                            reloadPkcException.addSuppressed(originalValidateSignatureException);
+                            LOGGER.debug(msg, reloadPkcException);
+                            listener.onResponse(AuthenticationResult.unsuccessful(msg, reloadPkcException));
+                            return;
```

Comment:
This can be a racing problem. If two thread tries to execute this reloading code block and Thread_1 finishes earlier, Thread_2 can possibly test `true` for this check because it could be comparing two "reloaded" JWKs, one reloaded by Thread_1 and the other by Thread_2. Authentication will fail in this case but it should not.

Reply:
I added ListenableFuture to fix this concurrency issue.